src/modules/aggregator/aggregator.c

```diff
@@ -0,0 +1,534 @@
+    }
+    if (!(o = aggregate_tojson (ag))) {
+        flux_log (h, LOG_ERR, "sink: aggregate_tojson failed");
+        return (-1);
+    }
+    if ((rc = kvs_put (h, ag->key, Jtostr (o))) < 0) {
+        flux_log_error (h, "sink: kvs_put");
+        goto out;
+    }
+    if ((rc = kvs_commit (h)) < 0)
+        flux_log_error (h, "sink: kvs_commit");
+out:
+    Jput (o);
+    return (rc);
+}
+
+static void aggregate_sink_again (flux_reactor_t *r, flux_watcher_t *w,
+                                  int revents, void *arg)
+{
+    struct aggregate *ag = arg;
+    flux_t h = ag->ctx->h;
+    if (aggregate_sink (h, ag) < 0)
+        aggregate_sink_abort (h, ag);
+    flux_watcher_destroy (w);
+    zhash_delete (ag->ctx->aggregates, ag->key);
+}
+
+
+/*
+ *   Push aggregate to kvs.
+ */
+static int aggregate_try_sink (flux_t h, struct aggregate *ag)
+{
+    if (aggregate_sink (h, ag) < 0) {
+        flux_watcher_t *w;
+        double t = ag->timeout;
+        if (t <= 1e-3)
+            t = .250;
+        flux_log (h, LOG_INFO, "sink: %s: retry  in %.3fs", ag->key, t);
+        /* On failure, retry just once, then abort */
+        w = flux_timer_watcher_create (flux_get_reactor (h),
+                                       t, 0.,
+                                       aggregate_sink_again,
+                                       (void *) ag);
+        flux_watcher_start (w);
+        return (-1);
+    }
+    zhash_delete (ag->ctx->aggregates, ag->key);
+    return (0);
+}
+
+/*
+ *  Flush aggregate `ag` -- forward entry upstream and destroy it locally.
+ */
+static int aggregate_flush (struct aggregate *ag)
+{
+    flux_t h = ag->ctx->h;
+    int rc;
+    assert (ag->ctx->rank != 0);
+    rc = aggregate_forward (h, ag);
+    zhash_delete (ag->ctx->aggregates, ag->key);
+    return (rc);
+}
+
+static void aggregate_destroy (struct aggregate *ag)
+{
+    struct aggregate_entry *ae = zlist_first (ag->entries);
+    while (ae) {
+        aggregate_entry_destroy (ae);
+        ae = zlist_next (ag->entries);
+    }
+    flux_watcher_destroy (ag->tw);
+    free (ag->key);
+    free (ag);
+}
+
+static void timer_cb (flux_reactor_t *r, flux_watcher_t *tw,
+                      int revents, void *arg)
+{
+    aggregate_flush (arg);
+}
+
+static void aggregate_timer_start (struct aggregator *ctx,
+                                   struct aggregate *ag,
+                                   double timeout)
+{
+    if (ctx->rank != 0) {
+        flux_t h = ctx->h;
+        flux_reactor_t *r = flux_get_reactor (h);
+        ag->tw = flux_timer_watcher_create (r, timeout, 0.,
+                                            timer_cb, (void *) ag);
+        if (ag->tw == NULL)
+            flux_log_error (h, "flux_timer_watcher_create");
+        flux_watcher_start (ag->tw);
+    }
+}
+
+static struct aggregate *
+    aggregate_create (struct aggregator *ctx, const char *key)
+{
+    flux_t h = ctx->h;
+
+    struct aggregate *ag = xzmalloc (sizeof (*ag));
```

Comment:
it occurs to me if `xzmalloc` is contraindicated for modules maybe we should wrap it in some sort of warning. Not sure if this is even possible though since we don't require modules to load `module.h` or anything... Just an idea to throw out there